gensim/models/fasttext.py

```diff
@@ -0,0 +1,374 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Author - Prakhar Pratyush (er.prakhar2b@gmail.com)
+
+import logging
+
+from gensim.models.word2vec import Word2Vec
+from gensim.models.ft_keyedvectors import FastTextKeyedVectors
+
+import numpy as np
+from numpy import exp, log, dot, zeros, outer, random, dtype, float32 as REAL,\
+    double, uint32, seterr, array, uint8, vstack, fromstring, sqrt, newaxis,\
+    ndarray, empty, sum as np_sum, prod, ones, ascontiguousarray, vstack, logaddexp
+
+from scipy.special import expit
+
+logger = logging.getLogger(__name__)
+
+
+"""
+TO-DO : description of FastText and the API
+"""
+
+MAX_WORDS_IN_BATCH = 10000
+
+def train_batch_sg(model, sentences, alpha, work=None):
+    """
+    Update skip-gram model by training on a sequence of sentences.
+
+    Each sentence is a list of string tokens, which are looked up in the model's
+    vocab dictionary. Called internally from `Word2Vec.train()`.
+
+    This is the non-optimized, Python version. If you have cython installed, gensim
+    will use the optimized version from word2vec_inner instead.
+
+    """
+    result = 0
+    for sentence in sentences:
+        word_vocabs = [model.wv.vocab[w] for w in sentence if w in model.wv.vocab and
+                       model.wv.vocab[w].sample_int > model.random.rand() * 2**32]
+        for pos, word in enumerate(word_vocabs):
+            reduced_window = model.random.randint(model.window)  # `b` in the original word2vec code
+
+            # subword_indices = []
+            # subword_indices += get_subwords(word)
+
+            # now go over all words from the (reduced) window, predicting each one in turn
+            start = max(0, pos - model.window + reduced_window)
+            for pos2, word2 in enumerate(word_vocabs[start:(pos + model.window + 1 - reduced_window)], start):
+                # don't train on the `word` itself
```

Comment:
I didn't change the comment here. I think it's correct as `don't train on word itself` means word is the target and rest words in windows are context (taken as input in word2vec code)